src/txn/txn.c

```diff
@@ -1016,10 +1017,54 @@ __txn_fixup_prepared_update(
                     WT_STAT_CONN_INCR(
                       session, txn_prepare_rollback_fix_hs_update_with_ckpt_reserved_txnid);
                 }
-            } else
+
+                /* Clear the history store flag. */
+                F_CLR(fix_upd, WT_UPDATE_HS);
+            } else {
+                flags = fix_upd->flags;
+                /* Set the flag to delete the value from the history store later. */
+                LF_SET(WT_UPDATE_TO_DELETE_FROM_HS);
+                /* Clear the history store flag and mark it to be deleted from the history store. */
+                LF_CLR(WT_UPDATE_HS);
+                /* We need to flip the flags atomically otherwise the compiler may reorder the code
+                 * to first clear the history store flag. That may cause the checkpoint to write
+                 * that value again to the history store if we are interrupted between. */
+                fix_upd->flags = flags;
+                /* We may not find a full update following the tombstone if it is obsolete. */
+                for (fix_upd = fix_upd->next; fix_upd != NULL; fix_upd = fix_upd->next)
+                    if (fix_upd->txnid != WT_TXN_ABORTED)
+                        break;
+                if (fix_upd != NULL) {
+                    WT_ASSERT(session, F_ISSET(fix_upd, WT_UPDATE_RESTORED_FROM_HS | WT_UPDATE_HS));
+                    flags = fix_upd->flags;
+                    /* Set the flag to delete the value from the history store later. */
+                    LF_SET(WT_UPDATE_TO_DELETE_FROM_HS);
+                    /* Clear the history store flag and mark it to be deleted from the history
+                     * store. */
+                    LF_CLR(WT_UPDATE_HS);
+                    /* We need to flip the flags atomically otherwise the compiler may reorder the
+                     * code to first clear the history store flag. That may cause the checkpoint to
+                     * write that value again to the history store if we are interrupted between. */
+                    fix_upd->flags = flags;
+                }
                 WT_STAT_CONN_INCR(session, txn_prepare_rollback_do_not_remove_hs_update);
-        } else
+            }
+        } else {
             WT_ERR(hs_cursor->remove(hs_cursor));
+            /* Clear the history store flag. */
```

Comment:
@agorrod @bvpvamsikrishna @kommiharibabu I think here we have another race. If we clear the flag after we remove the value from the history store and if we have a checkpoint starting and checkpoints this key in between, checkpoint will ignore the update restored because it still has the WT_UPDATE_HS flag on, which is wrong.

If we clear the WT_UPDATE_HS flag and then remove it from the history store, still a checkpoint can start to run and checkpoint the key in between, this time it will reinsert the update restored to the history store which hasn't been deleted and triggering out of order fix for that record. However, this out of order fix should not cause it to panic as it is the checkpoint that triggers it. May be that's acceptable?